axum-extra/src/multipart_builder.rs

```diff
@@ -0,0 +1,330 @@
+    /// ```
+    pub fn with_parts(parts: Vec<Part>) -> Self {
+        MultipartForm { parts }
+    }
+}
+
+impl IntoResponse for MultipartForm {
+    fn into_response(self) -> Response {
+        // see RFC2388 for details
+        let boundary = generate_boundary();
+        let mut headers = HeaderMap::new();
+        let mime_type: Mime = match format!("multipart/form-data; boundary={}", boundary).parse() {
+            Ok(m) => m,
+            // Realistically this should never happen unless the boundary generation code
+            // is modified, and that will be caught by unit tests
+            Err(_) => {
+                return (
+                    StatusCode::INTERNAL_SERVER_ERROR,
+                    "Invalid multipart boundary generated",
+                )
+                    .into_response()
+            }
+        };
+        // The use of unwrap is safe here because mime types are inherently string representable
+        headers.insert(header::CONTENT_TYPE, mime_type.to_string().parse().unwrap());
+        let mut serialized_form: Vec<u8> = Vec::new();
+        for part in self.parts {
+            // for each part, the boundary is preceded by two dashes
+            serialized_form.extend_from_slice(format!("--{}\r\n", boundary).as_bytes());
+            serialized_form.extend_from_slice(&part.serialize());
+        }
+        serialized_form.extend_from_slice(format!("--{}--", boundary).as_bytes());
+        (headers, serialized_form).into_response()
+    }
+}
+
+impl Default for MultipartForm {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+// Every part is expected to contain:
+// - a [Content-Disposition](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Disposition
+// header, where `Content-Disposition` is set to `form-data`, with a parameter of `name` that is set to
+// the name of the field in the form. In the below example, the name of the field is `user`:
+// ```
+// Content-Disposition: form-data; name="user"
+// ```
+// If the field contains a file, then the `filename` parameter may be set to the name of the file.
+// Handling for non-ascii field names is not done here, support for non-ascii characters may be encoded using
+// methodology described in RFC 2047.
+// - (optionally) a `Content-Type` header, which if not set, defaults to `text/plain`.
+// If the field contains a file, then the file should be identified with that file's MIME type (eg: `image/gif`).
+// If the `MIME` type is not known or specified, then the MIME type should be set to `application/octet-stream`.
+// - If the part does not conform to the default encoding, then the `Content-Transfer-Encoding` header may be supplied.
+// Valid settings for that header are: "base64", "quoted-printable", "8bit", "7bit", and "binary".
+/// A single part of a multipart form as defined by
+/// <https://www.w3.org/TR/html401/interact/forms.html#h-17.13.4>
+/// and RFC2388.
+#[derive(Debug)]
+pub struct Part {
```

Comment:
It's hard to tell where the regular comment ends and the doc comment starts. Maybe move the regular comment inside the definition and separate it from the first field's docs by an empty line?